File: drills/elixir/solutions/nc49_single_number__sum_of_uniques.exs
defmodule Solution do
  # Twice the sum of the distinct values counts every pair twice and the lone
  # value twice; subtracting the real total leaves the lone value. No bit
  # tricks, but it leans harder on the promise that everything else is a pair.
  def single_number(nums) do
    2 * (nums |> Enum.uniq() |> Enum.sum()) - Enum.sum(nums)
  end
end

File: drills/go/solutions/nc49_single_number__sum_of_uniques.go
package main

func singleNumber(nums []int) int {
	// Every value but one appears twice: twice the sum of the distinct
	// values, minus the sum of everything, is the one that appeared once.
	distinct := map[int]bool{}
	total, uniqueTotal := 0, 0
	for _, n := range nums {
		total += n
		if !distinct[n] {
			distinct[n] = true
			uniqueTotal += n
		}
	}
	return 2*uniqueTotal - total
}
